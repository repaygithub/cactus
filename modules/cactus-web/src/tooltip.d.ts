declare module '@reach/tooltip' {
  type EventFunction<EventType> = (event: EventType) => null

  interface trigger {
    'aria-describedby': string
    'data-reach-tooltip-trigger': string
    ref: React.MutableRefObject<HTMLDivElement | null>
    onMouseEnter: EventFunction<React.MouseEvent>
    onMouseMove: EventFunction<React.MouseEvent>
    onFocus: EventFunction<React.FocusEvent>
    onBlur: EventFunction<React.FocusEvent>
    onMouseLeave: EventFunction<React.MouseEvent>
    onKeyDown: EventFunction<React.KeyboardEvent>
    onMouseDown: EventFunction<React.MouseEvent>
  }

  interface tooltip {
    id: string
    triggerRect: DOMRect
    isVisible: boolean
  }

  type isVisible = boolean

  export const useTooltip: (params: object) => [trigger, tooltip, isVisible]
}
