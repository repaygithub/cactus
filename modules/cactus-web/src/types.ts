export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export type FieldOnChangeHandler<ValueType> = (fieldName: string, fieldValue: ValueType) => void

export type FieldOnFocusHandler = (fieldName: string) => void

export type FieldOnBlurHandler = (fieldName: string) => void
