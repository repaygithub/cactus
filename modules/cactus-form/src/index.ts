export { default as DependentField } from './DependentField'
export { default as FieldSpy } from './FieldSpy'
export { default as Field, validateRequired } from './Field'
