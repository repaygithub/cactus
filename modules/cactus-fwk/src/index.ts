export { useFeatureFlags, withFeatureFlags, FeatureFlag } from './featureFlags'
export { default } from './AppRoot'

export * from './types'
