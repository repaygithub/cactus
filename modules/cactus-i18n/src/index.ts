export {
  I18nProvider as default,
  I18nSection,
  I18nElement,
  I18nFormatted,
  I18nResource,
  I18nText,
} from './Components'
export { useI18nText, useI18nResource, useI18nSection, I18nContext } from './hooks'
export { default as BaseI18nController } from './BaseI18nController'
export * from './types'
