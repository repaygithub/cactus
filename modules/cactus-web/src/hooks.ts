// repay-scripts can't handle `index` files for some reason...
export { default as useControllableValue } from './hooks/useControllableValue'
export { default as useRefState } from './hooks/useRefState'
