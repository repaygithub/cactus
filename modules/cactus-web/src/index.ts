export { AccessibleField, useAccessibleField } from './AccessibleField/AccessibleField'
export { Accordion } from './Accordion/Accordion'
export { default as ActionBar } from './ActionBar/ActionBar'
export { default as ActionProvider, useAction, useActionBarItems } from './ActionBar/ActionProvider'
export { Alert } from './Alert/Alert'
export { Avatar } from './Avatar/Avatar'
export { Breadcrumb } from './Breadcrumb/Breadcrumb'
export { Button } from './Button/Button'
export { Box } from './Box/Box'
export { Card } from './Card/Card'
export { CheckBox } from './CheckBox/CheckBox'
export { CheckBoxField } from './CheckBoxField/CheckBoxField'
export { default as CheckBoxGroup } from './CheckBoxGroup/CheckBoxGroup'
export { ColorPicker } from './ColorPicker/ColorPicker'
export { DataGrid } from './DataGrid/DataGrid'
export { DateInput } from './DateInput/DateInput'
export { DateInputField } from './DateInputField/DateInputField'
export { default as Dimmer } from './Dimmer/Dimmer'
export { FieldWrapper } from './FieldWrapper/FieldWrapper'
export { FileInput } from './FileInput/FileInput'
export { FileInputField } from './FileInputField/FileInputField'
export { Flex } from './Flex/Flex'
export { default as Grid } from './Grid/Grid'
export { default as usePopup } from './helpers/usePopup'
export { IconButton } from './IconButton/IconButton'
export { Label } from './Label/Label'
export { default as Layout } from './Layout/Layout'
export { Link } from './Link/Link'
export { default as List } from './List/List'
export { MenuBar } from './MenuBar/MenuBar'
export { MenuButton } from './MenuButton/MenuButton'
export { Modal } from './Modal/Modal'
export { Notification } from './Notification/Notification'
export { Pagination } from './Pagination/Pagination'
export { Preview } from './Preview/Preview'
export { PrevNext } from './PrevNext/PrevNext'
export { RadioButton } from './RadioButton/RadioButton'
export { RadioButtonField } from './RadioButtonField/RadioButtonField'
export { default as RadioGroup } from './RadioGroup/RadioGroup'
export {
  ScreenSizeContext,
  ScreenSizeProvider,
  SIZES,
  useScreenSize,
} from './ScreenSizeProvider/ScreenSizeProvider'
export { Select } from './Select/Select'
export { SelectField } from './SelectField/SelectField'
export { Spinner } from './Spinner/Spinner'
export { SplitButton } from './SplitButton/SplitButton'
export { default as StatusMessage } from './StatusMessage/StatusMessage'
export { StepAvatar } from './StepAvatar/StepAvatar'
export { StyleProvider } from './StyleProvider/StyleProvider'
export { Tab, TabController, TabList, TabPanel } from './Tabs/Tabs'
export { Table } from './Table/Table'
export { Text, Span } from './Text/Text'
export { TextArea } from './TextArea/TextArea'
export { TextAreaField } from './TextAreaField/TextAreaField'
export { TextButton } from './TextButton/TextButton'
export { TextInput } from './TextInput/TextInput'
export { TextInputField } from './TextInputField/TextInputField'
export { Toggle } from './Toggle/Toggle'
export { ToggleField } from './ToggleField/ToggleField'
export { Tooltip } from './Tooltip/Tooltip'
export { default as useId } from './helpers/useId'
export { ConfirmModal } from './ConfirmModal/ConfirmModal'
export { Tag } from './Tag/Tag'
export { Footer } from './Footer/Footer'
export { default as BrandBar } from './BrandBar/BrandBar'
export { Divider } from './Divider/Divider'
export { Header } from './Header/Header'
