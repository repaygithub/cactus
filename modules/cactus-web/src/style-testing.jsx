import { screenSizes } from '@repay/cactus-theme'
import { omit } from 'lodash'
import React from 'react'
import styled from 'styled-components'
import { margin } from 'styled-system'

import { SIZES, useScreenSize } from './ScreenSizeProvider/ScreenSizeProvider'

const inc = (s) => s + 1
const sum = (s, n) => s + n

export default function TestBed({ Component, repeat = 1000, ...rest }) {
  const [current, setNext] = React.useState(0)
  const ref = React.useRef(Date.now())
  React.useEffect(() => {
    if (current < repeat) {
      setTimeout(() => setNext(inc), 0)
    } else {
      const t = times.get(Component)
      t.push(Date.now() - ref.current)
      const total = t.reduce(sum, 0) / t.length
      console.log('TOTAL TIME', Component.displayName, ':', total / 1000)
    }
  })
  return React.createElement(Component, rest, current)
}

export const TestBed2 = ({ Component, level = 2, ...rest }) => {
  if (level > 0) {
    rest.level = level - 1
    rest.Component = Component
    const children = []
    for (let i = 0; i < 50; i++) {
      rest.key = i
      children.push(React.createElement(TestBed2, rest))
    }
    return <>{children}</>
  }
  return React.createElement(Component, rest, level)
}

export const Control = styled.div(margin)

// TODO In the real thing these would need to be generated dynamically, but this should be fine for a performance test.
const queries = [
  '@media screen and (min-width: 768px)',
  '@media screen and (min-width: 1024px)',
  '@media screen and (min-width: 1200px)',
  '@media screen and (min-width: 1440px)',
]

const sizes = [SIZES.small, SIZES.medium, SIZES.large, SIZES.extraLarge]

const useDefaultParserAndHook = (props) => {
  const parsedStyles = margin(props)
  const size = useScreenSize()
  const baseStyles = omit(parsedStyles, queries)
  let i = 0
  while (sizes[i] <= size) {
    Object.assign(baseStyles, parsedStyles[queries[i++]])
  }
  return { style: props.style ? { ...props.style, ...baseStyles } : baseStyles }
}

export const WithDefaultParserAndHook = styled.div.attrs(useDefaultParserAndHook)`
  width: 50px;
  height: 50px;
  background: red;
`

const config = margin.config
const useCustomParserAndHook = (props) => {
  const currentSize = useScreenSize()
  const { style: styleProp, theme } = props
  let style = styleProp
  let sizeKeys = false
  for (const key in props) {
    const processor = config[key]
    if (!processor) continue
    const scale = theme[processor.scale] || processor.defaults
    const rawStyle = props[key]

    const merge = (value) => {
      if (value !== null && value !== undefined) {
        const processed = processor(value, scale, props)
        if (processed) {
          if (style === styleProp) {
            style = Object.assign({}, style, processed)
          } else {
            Object.assign(style, processed)
          }
          return true
        }
      }
    }
    if (typeof rawStyle === 'object' && rawStyle !== null) {
      // Returns an array of matching sizes, e.g if current screen size
      // is medium, returns `['tiny', 'small', 'medium']` (ascending order).
      if (!sizeKeys) {
        sizeKeys = sizes.reduce(
          (keys, size) => {
            if (size <= currentSize) keys.push(size.size)
            return keys
          },
          ['tiny']
        )
        sizeKeys.reverse()
      }

      if (Array.isArray(rawStyle)) {
        for (let i = Math.min(sizeKeys.length, rawStyle.length) - 1; i >= 0; i--) {
          if (merge(rawStyle[i])) break
        }
      } else {
        for (const sizeKey of sizeKeys) {
          if (merge(rawStyle[sizeKey])) break
        }
      }
    } else {
      merge(rawStyle)
    }
  }
  return style ? { style } : undefined
}

export const WithCustomParserAndHook = styled.div.attrs(useCustomParserAndHook)`
  width: 50px;
  height: 50px;
  background: yellow;
`

const useDefaultParserAndTheme = (props) => {
  const parsedStyles = margin(props)
  const baseStyles = omit(parsedStyles, queries)
  for (const size of sizes) {
    const q = props.theme.mediaQueries[size.size]
    if (q.matches) {
      Object.assign(baseStyles, parsedStyles[`@media ${q.media}`])
    }
  }
  return { style: props.style ? { ...props.style, ...baseStyles } : baseStyles }
}

export const WithDefaultParserAndTheme = styled.div.attrs(useDefaultParserAndTheme)`
  width: 50px;
  height: 50px;
  background: blue;
`

const useCustomParserAndTheme = (props) => {
  const { style: styleProp, theme } = props
  let style = styleProp
  let sizeKeys = false
  for (const key in props) {
    const processor = config[key]
    if (!processor) continue
    const scale = theme[processor.scale] || processor.defaults
    const rawStyle = props[key]

    const merge = (value) => {
      if (value !== null && value !== undefined) {
        const processed = processor(value, scale, props)
        if (processed) {
          if (style === styleProp) {
            style = Object.assign({}, style, processed)
          } else {
            Object.assign(style, processed)
          }
          return true
        }
      }
    }
    if (typeof rawStyle === 'object' && rawStyle !== null) {
      // Returns an array of matching sizes, e.g if current screen size
      // is medium, returns `['tiny', 'small', 'medium']` (ascending order).
      if (!sizeKeys) {
        sizeKeys = []
        for (const size of screenSizes) {
          if (theme.mediaQueries[size].matches) sizeKeys.push(size)
        }
        sizeKeys.reverse()
      }

      if (Array.isArray(rawStyle)) {
        for (let i = Math.min(sizeKeys.length, rawStyle.length) - 1; i >= 0; i--) {
          if (merge(rawStyle[i])) break
        }
      } else {
        for (const sizeKey of sizeKeys) {
          if (merge(rawStyle[sizeKey])) break
        }
      }
    } else {
      merge(rawStyle)
    }
  }
  return style ? { style } : undefined
}

export const WithCustomParserAndTheme = styled.div.attrs(useCustomParserAndTheme)`
  width: 50px;
  height: 50px;
  background: green;
`

// Custom parser makes two assumptions about the style processors:
// 1. Must return simple style object, i.e. all the keys must be CSS properties;
//    can't return something like `{ '& > *': { margin: 5 } }` to set child styles.
// 2. Must return all styles for a given input; can't rely on screen size cascading.
// (I'm actually not even sure if it's possible to violate #2, but maybe in theory...)
// As far as I know, all the built-in `styled-system` processors meet these assumptions.

const times = new Map()
times.set(Control, [])
times.set(WithCustomParserAndHook, [])
times.set(WithCustomParserAndTheme, [])
times.set(WithDefaultParserAndHook, [])
times.set(WithDefaultParserAndTheme, [])
