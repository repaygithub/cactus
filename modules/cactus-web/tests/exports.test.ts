import * as components from '../src/index'

describe('library exports', () => {
  test('should match previous exported names', () => {
    // removed or renamed exports are a breaking change
    expect(Object.keys(components)).toMatchSnapshot()
  })
})
