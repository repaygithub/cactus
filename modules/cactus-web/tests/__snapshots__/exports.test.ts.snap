// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`library exports should match previous exported names 1`] = `
Array [
  "AccessibleField",
  "Accordion",
  "ActionBar",
  "ActionProvider",
  "Alert",
  "Avatar",
  "Box",
  "BrandBar",
  "Breadcrumb",
  "Button",
  "Calendar",
  "Card",
  "CheckBox",
  "CheckBoxCard",
  "CheckBoxField",
  "CheckBoxGroup",
  "ColorPicker",
  "ConfirmModal",
  "DataGrid",
  "DateInput",
  "DateInputField",
  "Dimmer",
  "Divider",
  "FieldWrapper",
  "FileInput",
  "FileInputField",
  "Flex",
  "Footer",
  "Grid",
  "Header",
  "IconButton",
  "Label",
  "Layout",
  "Link",
  "List",
  "MenuBar",
  "MenuButton",
  "Modal",
  "Notification",
  "Pagination",
  "PrevNext",
  "Preview",
  "RadioButton",
  "RadioButtonField",
  "RadioCard",
  "RadioGroup",
  "SIZES",
  "ScreenSizeContext",
  "ScreenSizeProvider",
  "Select",
  "SelectField",
  "Spinner",
  "SplitButton",
  "StatusMessage",
  "StepAvatar",
  "StyleProvider",
  "Tab",
  "TabController",
  "TabList",
  "TabPanel",
  "Table",
  "Tag",
  "Text",
  "TextArea",
  "TextAreaField",
  "TextButton",
  "TextInput",
  "TextInputField",
  "Toggle",
  "ToggleField",
  "Tooltip",
  "trapScroll",
  "useAccessibleField",
  "useAction",
  "useActionBarItems",
  "useId",
  "useLayout",
  "usePopup",
  "useScreenSize",
  "useScrollTrap",
]
`;
